### PR TITLE
HW #2 (SOLID & GRASP)

### PR DESCRIPTION
**Pull request by PavloVlasiuk**
Changes:
- split email service on two services: one that sends emails with external library mailer and another that works with email entity and sends email messages
- rewrite TaskScheduleService method to execute the logic of getting rate and sending emails only
- move email exceptions to email domain
- rewrite tests because changes has impact on structure of services

### CODE REVIEW

**Comment by Svitlana-Marchenko** [comment]
Your code looks great.
It's good that you didn't forget about the tests and changed them and added new due to code changes.
I like your commit "a83c345d" where you changed several classes to interfaces, as it improved the flexibility.

File: src/email/services/email.service.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+  async subscribe({ email }: SubscribeEmailDto): Promise<void> {
+    const alreadySubscribed = await this.emailRepository.find({ email });
+
+    if (alreadySubscribed) throw new AlreadySubscribedException();
+
+    await this.emailRepository.create({ email });
+  }
+
+  async sendCurrentRate({ rate }: IExchangeRate): Promise<void> {
```
Comment (bizkitnu) [inline src/email/services/email.service.ts:26]: I suggest you to rename this function just to sendRate

Comment (trndo) [inline src/email/services/email.service.ts]: I suggest to rename this to sendRate

File: src/rate/interfaces/get-exchange-rate.interface.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+export interface IGetExchangeRate {
+  r030: number;
+  txt: string;
+  rate: number;
+  cc: string;
+  exchangedate: string;
+}
```
Comment (trndo) [inline src/rate/interfaces/get-exchange-rate.interface.ts:7]: Can you explain pls what do you solve with this Interface in your code ?

File: src/email/services/email.service.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import { IExchangeRate } from '../../rate/interfaces';
+import { SubscribeEmailDto } from '../dtos/subscribe-email.dto';
+import { AlreadySubscribedException } from '../exceptions';
+import { ISendCurrentRateContext } from '../interfaces/send-email-options.interface';
+
+@Injectable()
+export class EmailService {
+  constructor(
+    private readonly emailRepository: EmailRepository,
+    private readonly nodeMailerService: NodeMailerService,
```
Comment (trndo) [inline src/email/services/email.service.ts]: What will you do here if the requirements change and you need to use for example MailChimp client for email sending or other one, not NodeMailer ?

File: src/rate/interfaces/get-exchange-rate.interface.ts (hunk 2 above)
Comment (PavloVlasiuk) [reply src/rate/interfaces/get-exchange-rate.interface.ts:7]: This interface is used as data contract for function that calls NBU API to obtain information about exchange rate. In other words I tell the function to return data exactly in this form

File: src/email/services/email.service.ts (hunk 3 above)
Comment (PavloVlasiuk) [reply src/email/services/email.service.ts]: If I need to add one more client for sending email, I would use Strategy pattern to use different implementation of email sending logic. For now, I should create an interface for the service that implements email sending logic to be able to create new implementations of email clients based on that interface.  And I think that name nodeMailerService can be replaced with more common name like MailingService.

**Review by bizkitnu — APPROVED** [review]

**Review by trndo — APPROVED** [review]

**Pull request merged**